chiselFrontend/src/main/scala/chisel3/core/Data.scala

```diff
@@ -236,16 +273,38 @@ abstract class Data extends HasId {
     * @note does NOT check bit widths, may drop bits during assignment
     * @note what fromBits assigs to must have known widths
     */
-  def fromBits(that: Bits): this.type = macro CompileOptionsTransform.thatArg
-
-  def do_fromBits(that: Bits)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): this.type
+  def fromBits(that: Bits)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): this.type = {
+    val output = Wire(chiselCloneType).asInstanceOf[this.type]
+    output.connectFromBits(that)
+    output
+  }
 
   /** Packs the value of this object as plain Bits.
     *
     * This performs the inverse operation of fromBits(Bits).
     */
   @deprecated("Best alternative, .asUInt()", "chisel3")
-  def toBits(): UInt = SeqUtils.do_asUInt(this.flatten)(DeprecatedSourceInfo)
+  def toBits(): UInt = do_asUInt(DeprecatedSourceInfo)
+
+  /** Does a reinterpret cast of the bits in this node into the format that provides.
+    * Returns a new Wire of that type. Does not modify existing nodes.
+    *
+    * x.asTypeOf(that) performs the inverse operation of x = that.toBits.
+    *
+    * @note bit widths are NOT checked, may pad or drop bits from input
+    * @note that should have known widths
+    */
+  def asTypeOf[T <: Data](that: T): T = macro CompileOptionsTransform.thatArg
+
+  def do_asTypeOf[T <: Data](that: T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T = {
+    val thatCloned = Wire(that.chiselCloneType)
+    thatCloned.connectFromBits(asUInt())
```

Comment:
this is just my opinion but I generally prefer to use `this.` when invoking methods on the object at hand. That `asUInt()` just looks like some magical functional call rather than calling something on `this`